bodhi/tests/server/scripts/test_approve_testing.py

```diff
@@ -822,6 +822,29 @@ def test_autotime_update_with_autokarma_met_karma_and_time_requirements_get_push
         assert cmnts.count() == 1
         assert cmnts[0].text == 'This update has been submitted for stable by bodhi. '
 
+    def test_autotime_update_no_autokarma_negative_karma_not_pushed(self):
+        """
+        Ensure that an autotime update which negative karma does not get pushed.
+        """
+        update = self.db.query(models.Update).all()[0]
+        update.autokarma = False
+        update.autotime = True
+        update.request = None
+        update.stable_karma = 1
+        update.stable_days = 0
+        update.date_testing = datetime.utcnow() - timedelta(days=8)
+        update.status = models.UpdateStatus.testing
+        update.comment(self.db, u'Broken', author=u'luke', karma=-1)
+        with fml_testing.mock_sends(api.Message):
+            self.db.commit()
+
+        with patch('bodhi.server.scripts.approve_testing.initialize_db'):
+            with patch('bodhi.server.scripts.approve_testing.get_appsettings', return_value=''):
+                approve_testing.main(['nosetests', 'some_config.ini'])
```

Comment:
For me, the tests hang here. Running it in a debugger shows me it tries to send this message (ellipses by me):

```
ipdb> m
UpdateRequirementsMetStableV1(id='9ada842e-04cd-485d-b44c-cda985ebec8c', topic='bodhi.update.requirements_met.stable', body={...})
ipdb> pprint.pprint(m.body)
{'update': {'alias': 'FEDORA-2019-a3bbe1a8f2',
            'autokarma': False,
            'autotime': False,
            'bugs': [{'bug_id': 12345,
                      'feedback': [],
                      'parent': False,
                      'security': False,
                      'title': None}],
            'builds': [{'epoch': 0,
                        'nvr': 'bodhi-2.0-1.fc17',
                        'release_id': 1,
                        'signed': True,
                        'type': 'rpm'}],
            'close_bugs': True,
            'comments': [{'bug_feedback': [],
                          'id': 1,
                          'karma': 1,
                          'karma_critpath': 0,
                          'testcase_feedback': [],
                          'text': 'wow. amaze.',
                          'timestamp': '2019-08-13 13:08:58',
                          'update_id': 1,
                          'user': {'avatar': None,
                                   'email': None,
                                   'groups': [{'name': 'packager'}],
                                   'id': 1,
                                   'name': 'guest',
                                   'openid': None},
                          'user_id': 1},
                         {'bug_feedback': [],
                          'id': 2,
                          'karma': 0,
                          'karma_critpath': 0,
                          'testcase_feedback': [],
                          'text': 'srsly.  pretty good.',
                          'timestamp': '2019-08-13 13:08:58',
                          'update_id': 1,
                          'user': {'avatar': None,
                                   'email': None,
                                   'groups': [],
                                   'id': 2,
                                   'name': 'anonymous',
                                   'openid': None},
                          'user_id': 2},
                         {'bug_feedback': [],
                          'id': 3,
                          'karma': -1,
                          'karma_critpath': 0,
                          'testcase_feedback': [],
                          'text': 'Broken',
                          'timestamp': '2019-08-13 13:08:59',
                          'update_id': 1,
                          'user': {'avatar': None,
                                   'email': None,
                                   'groups': [],
                                   'id': 3,
                                   'name': 'luke',
                                   'openid': None},
                          'user_id': 3},
                         {'bug_feedback': [],
                          'id': 4,
                          'karma': 0,
                          'karma_critpath': 0,
                          'testcase_feedback': [],
                          'text': 'Bodhi is disabling automatic push to stable '
                                  'due to negative karma. The maintainer may '
                                  'push manually if they determine that the '
                                  'issue is not severe.',
                          'timestamp': '2019-08-13 13:08:59',
                          'update_id': 1,
                          'user': {'avatar': None,
                                   'email': None,
                                   'groups': [],
                                   'id': 4,
                                   'name': 'bodhi',
                                   'openid': None},
                          'user_id': 4},
                         {'bug_feedback': [],
                          'id': 5,
                          'karma': 0,
                          'karma_critpath': 0,
                          'testcase_feedback': [],
                          'text': 'This update can be pushed to stable now if '
                                  'the maintainer wishes',
                          'timestamp': '2019-08-13 13:08:59',
                          'update_id': 1,
                          'user': {'avatar': None,
                                   'email': None,
                                   'groups': [],
                                   'id': 4,
                                   'name': 'bodhi',
                                   'openid': None},
                          'user_id': 4}],
            'compose': None,
            'content_type': 'rpm',
            'critpath': False,
            'date_approved': None,
            'date_modified': None,
            'date_pushed': None,
            'date_stable': None,
            'date_submitted': '1984-11-02 00:00:00',
            'date_testing': '2019-08-05 13:08:59',
            'display_name': '',
            'from_tag': None,
            'karma': 0,
            'locked': False,
            'meets_testing_requirements': True,
            'notes': 'Useful details!',
            'pushed': False,
            'release': {'branch': 'f17',
                        'candidate_tag': 'f17-updates-candidate',
                        'composed_by_bodhi': True,
                        'composes': [],
                        'create_automatic_updates': True,
                        'dist_tag': 'f17',
                        'id_prefix': 'FEDORA',
                        'long_name': 'Fedora 17',
                        'mail_template': 'fedora_errata_template',
                        'name': 'F17',
                        'override_tag': 'f17-override',
                        'package_manager': 'unspecified',
                        'pending_signing_tag': 'f17-updates-testing-signing',
                        'pending_stable_tag': 'f17-updates-pending',
                        'pending_testing_tag': 'f17-updates-testing-pending',
                        'stable_tag': 'f17-updates',
                        'state': 'current',
                        'testing_repository': None,
                        'testing_tag': 'f17-updates-testing',
                        'version': '17'},
            'request': None,
            'require_bugs': False,
            'require_testcases': False,
            'requirements': 'rpmlint',
            'severity': 'medium',
            'stable_days': 0,
            'stable_karma': 1,
            'status': 'testing',
            'suggest': 'unspecified',
            'test_cases': [{'name': 'Wat', 'package_id': 1}],
            'test_gating_status': None,
            'title': 'bodhi-2.0-1.fc17',
            'type': 'bugfix',
            'unstable_karma': -3,
            'updateid': 'FEDORA-2019-a3bbe1a8f2',
            'url': 'http://0.0.0.0:6543/updates/FEDORA-2019-a3bbe1a8f2',
            'user': {'avatar': None,
                     'email': None,
                     'groups': [{'name': 'packager'}],
                     'id': 1,
                     'name': 'guest',
                     'openid': None}}}
ipdb>
```

Is this expected?

Reply:
> For me, the tests hang here. Running it in a debugger shows me it tries to send this message (ellipses by me):
> 
> ```
> ipdb> m
> UpdateRequirementsMetStableV1(id='9ada842e-04cd-485d-b44c-cda985ebec8c', topic='bodhi.update.requirements_met.stable', body={...})
> ipdb> pprint.pprint(m.body)
> {'update': {'alias': 'FEDORA-2019-a3bbe1a8f2',
>             'autokarma': False,
>             'autotime': False,
>             'bugs': [{'bug_id': 12345,
>                       'feedback': [],
>                       'parent': False,
>                       'security': False,
>                       'title': None}],
>             'builds': [{'epoch': 0,
>                         'nvr': 'bodhi-2.0-1.fc17',
>                         'release_id': 1,
>                         'signed': True,
>                         'type': 'rpm'}],
>             'close_bugs': True,
>             'comments': [{'bug_feedback': [],
>                           'id': 1,
>                           'karma': 1,
>                           'karma_critpath': 0,
>                           'testcase_feedback': [],
>                           'text': 'wow. amaze.',
>                           'timestamp': '2019-08-13 13:08:58',
>                           'update_id': 1,
>                           'user': {'avatar': None,
>                                    'email': None,
>                                    'groups': [{'name': 'packager'}],
>                                    'id': 1,
>                                    'name': 'guest',
>                                    'openid': None},
>                           'user_id': 1},
>                          {'bug_feedback': [],
>                           'id': 2,
>                           'karma': 0,
>                           'karma_critpath': 0,
>                           'testcase_feedback': [],
>                           'text': 'srsly.  pretty good.',
>                           'timestamp': '2019-08-13 13:08:58',
>                           'update_id': 1,
>                           'user': {'avatar': None,
>                                    'email': None,
>                                    'groups': [],
>                                    'id': 2,
>                                    'name': 'anonymous',
>                                    'openid': None},
>                           'user_id': 2},
>                          {'bug_feedback': [],
>                           'id': 3,
>                           'karma': -1,
>                           'karma_critpath': 0,
>                           'testcase_feedback': [],
>                           'text': 'Broken',
>                           'timestamp': '2019-08-13 13:08:59',
>                           'update_id': 1,
>                           'user': {'avatar': None,
>                                    'email': None,
>                                    'groups': [],
>                                    'id': 3,
>                                    'name': 'luke',
>                                    'openid': None},
>                           'user_id': 3},
>                          {'bug_feedback': [],
>                           'id': 4,
>                           'karma': 0,
>                           'karma_critpath': 0,
>                           'testcase_feedback': [],
>                           'text': 'Bodhi is disabling automatic push to stable '
>                                   'due to negative karma. The maintainer may '
>                                   'push manually if they determine that the '
>                                   'issue is not severe.',
>                           'timestamp': '2019-08-13 13:08:59',
>                           'update_id': 1,
>                           'user': {'avatar': None,
>                                    'email': None,
>                                    'groups': [],
>                                    'id': 4,
>                                    'name': 'bodhi',
>                                    'openid': None},
>                           'user_id': 4},
>                          {'bug_feedback': [],
>                           'id': 5,
>                           'karma': 0,
>                           'karma_critpath': 0,
>                           'testcase_feedback': [],
>                           'text': 'This update can be pushed to stable now if '
>                                   'the maintainer wishes',
>                           'timestamp': '2019-08-13 13:08:59',
>                           'update_id': 1,
>                           'user': {'avatar': None,
>                                    'email': None,
>                                    'groups': [],
>                                    'id': 4,
>                                    'name': 'bodhi',
>                                    'openid': None},
>                           'user_id': 4}],
>             'compose': None,
>             'content_type': 'rpm',
>             'critpath': False,
>             'date_approved': None,
>             'date_modified': None,
>             'date_pushed': None,
>             'date_stable': None,
>             'date_submitted': '1984-11-02 00:00:00',
>             'date_testing': '2019-08-05 13:08:59',
>             'display_name': '',
>             'from_tag': None,
>             'karma': 0,
>             'locked': False,
>             'meets_testing_requirements': True,
>             'notes': 'Useful details!',
>             'pushed': False,
>             'release': {'branch': 'f17',
>                         'candidate_tag': 'f17-updates-candidate',
>                         'composed_by_bodhi': True,
>                         'composes': [],
>                         'create_automatic_updates': True,
>                         'dist_tag': 'f17',
>                         'id_prefix': 'FEDORA',
>                         'long_name': 'Fedora 17',
>                         'mail_template': 'fedora_errata_template',
>                         'name': 'F17',
>                         'override_tag': 'f17-override',
>                         'package_manager': 'unspecified',
>                         'pending_signing_tag': 'f17-updates-testing-signing',
>                         'pending_stable_tag': 'f17-updates-pending',
>                         'pending_testing_tag': 'f17-updates-testing-pending',
>                         'stable_tag': 'f17-updates',
>                         'state': 'current',
>                         'testing_repository': None,
>                         'testing_tag': 'f17-updates-testing',
>                         'version': '17'},
>             'request': None,
>             'require_bugs': False,
>             'require_testcases': False,
>             'requirements': 'rpmlint',
>             'severity': 'medium',
>             'stable_days': 0,
>             'stable_karma': 1,
>             'status': 'testing',
>             'suggest': 'unspecified',
>             'test_cases': [{'name': 'Wat', 'package_id': 1}],
>             'test_gating_status': None,
>             'title': 'bodhi-2.0-1.fc17',
>             'type': 'bugfix',
>             'unstable_karma': -3,
>             'updateid': 'FEDORA-2019-a3bbe1a8f2',
>             'url': 'http://0.0.0.0:6543/updates/FEDORA-2019-a3bbe1a8f2',
>             'user': {'avatar': None,
>                      'email': None,
>                      'groups': [{'name': 'packager'}],
>                      'id': 1,
>                      'name': 'guest',
>                      'openid': None}}}
> ipdb>
> ```
> 
> Is this expected?

Ha good catch, I always forget about these :-). Yes it is expected since the update requirements are met, negative karma will just disable the automatic push either via karma or via time.

Reply:
Thanks!